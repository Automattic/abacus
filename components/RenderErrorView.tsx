import React from 'react'

import type { RenderError } from '@/components/RenderErrorBoundary'

type Props = {
  renderError: RenderError
}

const RenderErrorView = (props: Props) => {
  return (
    <>
      <strong>Oops!</strong>
      <p>{props.renderError.error.message}</p>
      <p>An error occurred. If error persists, please contact the Experiments Platform team.</p>
      <button onClick={props.renderError.clear}>Clear</button>
    </>
  )
}

export default RenderErrorView
