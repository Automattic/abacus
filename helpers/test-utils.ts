import mediaQuery from 'css-mediaquery'

/**
 * Create a `matchMedia` function that will match a query based on the specified
 * width.
 *
 *     const initialJsDomWindowInnerWidth = window.innerWidth
 *     afterEach(() => {
 *       // Reset back to initial width for tests that don't explicitly set the width
 *       // themselves.
 *       window.matchMedia = createMatchMedia(initialJsDomWindowInnerWidth)
 *     })
 *
 *     test('...', () => {
 *       window.matchMedia = createMatchMedia(600)
 *       ...
 *     })
 *
 * @param width - The width of the window to simulate.
 */
function createMatchMedia(width: number) {
  return (query: string) => ({
    matches: mediaQuery.match(query, { width }),
    media: query,
    addEventListener: jest.fn(),
    addListener: jest.fn(),
    dispatchEvent: jest.fn(),
    onchange: jest.fn(),
    removeEventListener: jest.fn(),
    removeListener: jest.fn(),
  })
}

export { createMatchMedia }
