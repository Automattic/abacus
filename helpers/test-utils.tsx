import { Queries, render as actualRender, RenderOptions } from '@testing-library/react'
import mediaQuery from 'css-mediaquery'
import React from 'react'

import ThemeProvider from '@/styles/ThemeProvider'

/**
 * A wrapped unit-test react-renderer, useful for adding React Contexts globally.
 */
export const render: typeof actualRender = <Q extends Queries>(ui: React.ReactElement, options?: RenderOptions<Q>) =>
  actualRender((<ThemeProvider>{ui}</ThemeProvider>) as React.ReactElement, options) as ReturnType<typeof actualRender>

/**
 * Create a `matchMedia` function that will match a query based on the specified
 * width.
 *
 *     const initialJsDomWindowInnerWidth = window.innerWidth
 *     afterEach(() => {
 *       // Reset back to initial width for tests that don't explicitly set the width
 *       // themselves.
 *       window.matchMedia = createMatchMedia(initialJsDomWindowInnerWidth)
 *     })
 *
 *     test('...', () => {
 *       window.matchMedia = createMatchMedia(600)
 *       ...
 *     })
 *
 * @param width - The width of the window to simulate.
 */
export function createMatchMedia(width: number) {
  return (query: string) => ({
    matches: mediaQuery.match(query, { width }),
    media: query,
    addEventListener: jest.fn(),
    addListener: jest.fn(),
    dispatchEvent: jest.fn(),
    onchange: jest.fn(),
    removeEventListener: jest.fn(),
    removeListener: jest.fn(),
  })
}
