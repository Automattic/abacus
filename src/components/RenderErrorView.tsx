import React from 'react'

import type { RenderError } from '@/components/RenderErrorBoundary'

export default function (props: { renderError: RenderError }): JSX.Element {
  return (
    <>
      <strong>Oops!</strong>
      <p>{props.renderError.error.message}</p>
      <p>An error occurred. If error persists, please contact the Experiments Platform team.</p>
      <button onClick={props.renderError.clear}>Clear</button>
    </>
  )
}
