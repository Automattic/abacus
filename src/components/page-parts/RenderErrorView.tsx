import {
  Button,
  createStyles,
  Dialog,
  DialogActions,
  DialogContent,
  DialogTitle,
  makeStyles,
  Theme,
  Typography,
} from '@material-ui/core'
import React from 'react'

import type { RenderError } from 'src/components/page-parts/RenderErrorBoundary'

const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    root: {},
    pre: {
      overflow: 'scroll',
      maxHeight: '10rem',
      background: '#f5f5f5',
      padding: theme.spacing(2),
      marginTop: 0,
    },
  }),
)

export default function (props: { renderError: RenderError }): JSX.Element {
  const classes = useStyles()
  return (
    <div className={classes.root}>
      <Dialog open={true} maxWidth={false}>
        <DialogTitle disableTypography>
          <Typography variant='h5'>Oops! Something went wrong...</Typography>
        </DialogTitle>
        <DialogContent>
          <pre className={classes.pre}>{props.renderError.error.stack || props.renderError.error.message}</pre>
          <Typography>If error persists, please contact the Experiments Platform team.</Typography>
        </DialogContent>
        <DialogActions>
          <Button component={'a'} href='/'>
            Go to home
          </Button>
          <Button onClick={props.renderError.clear} color='primary'>
            Try again
          </Button>
        </DialogActions>
      </Dialog>
    </div>
  )
}
