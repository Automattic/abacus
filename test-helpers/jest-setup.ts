import { format } from 'util'

// Throw errors on any console logging.
// Adapted from https://github.com/facebook/jest/issues/6121#issuecomment-529591574
;['error', 'info', 'log', 'warn'].forEach((funcName) => {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  ;(global.console as any)[funcName] = function (message?: any, ...optionalParams: any[]) {
    throw new Error(format(message, optionalParams))
  }
})
