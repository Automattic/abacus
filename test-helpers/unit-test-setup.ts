/* eslint-disable @typescript-eslint/ban-ts-ignore */
// @ts-ignore
window.URL.createObjectURL = jest.fn()

// @ts-ignore
window.HTMLCanvasElement.prototype.getContext = jest.fn()
