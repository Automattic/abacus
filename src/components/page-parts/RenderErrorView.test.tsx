/* eslint-disable @typescript-eslint/no-unsafe-assignment */
import React from 'react'

import { render } from 'src/test-helpers/test-utils'

import RenderErrorView from './RenderErrorView'

test('renders error stack when available', () => {
  const renderError = {
    clear: jest.fn(),
    error: Error('The error message.'),
    info: expect.any(Object),
  }
  const { container } = render(<RenderErrorView renderError={renderError} />)

  expect(container).toMatchSnapshot()
})

test(`renders error message when stack isn't available`, () => {
  const renderError = {
    clear: jest.fn(),
    error: { message: 'the error message' } as Error,
    info: expect.any(Object),
  }
  const { container } = render(<RenderErrorView renderError={renderError} />)

  expect(container).toMatchSnapshot()
})
