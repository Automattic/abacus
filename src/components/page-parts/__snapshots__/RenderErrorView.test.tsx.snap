// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders error message when stack isn't available 1`] = `
<div
  aria-hidden="true"
>
  <div
    class="makeStyles-root-4"
  />
</div>
`;

exports[`renders error stack when available 1`] = `
<div
  aria-hidden="true"
>
  <div
    class="makeStyles-root-1"
  />
</div>
`;
